implementations/rust/ockam/ockam_transport_uds/src/workers/sender.rs

```diff
@@ -0,0 +1,322 @@
+
+use crate::router::UdsRouterHandle;
+
+use super::UdsRecvProcessor;
+
+/// Provides the transmit and Socket Addr of a UDS connection
+#[derive(Debug)]
+pub(crate) struct WorkerPair {
+    paths: Vec<String>,
+    peer: SocketAddr,
+    tx_addr: Address,
+}
+
+impl WorkerPair {
+    /// Returns a reference to the peers pathnames
+    pub fn paths(&self) -> &[String] {
+        &self.paths
+    }
+
+    /// Return a reference to the peers SocketAddr
+    pub fn peer(&self) -> &SocketAddr {
+        &self.peer
+    }
+
+    /// Return a clone of the transmit
+    pub fn tx_addr(&self) -> Address {
+        self.tx_addr.clone()
+    }
+}
+
+#[derive(Serialize, Deserialize, Message, Clone)]
+pub(crate) enum UdsSendWorkerMsg {
+    ConnectionClosed,
+}
+
+pub(crate) struct UdsSendWorker {
+    router_handle: UdsRouterHandle,
+    rx: Option<OwnedReadHalf>,
+    tx: Option<OwnedWriteHalf>,
+    peer: SocketAddr,
+    internal_addr: Address,
+    rx_addr: Option<Address>,
+}
+
+impl UdsSendWorker {
+    /// Create a new [`UdsSendWorker`]
+    fn new(
+        router_handle: UdsRouterHandle,
+        stream: Option<UnixStream>,
+        peer: SocketAddr,
+        internal_addr: Address,
+    ) -> Self {
+        let (rx, tx) = match stream {
+            Some(s) => {
+                let (rx, tx) = s.into_split();
+                (Some(rx), Some(tx))
+            }
+            None => (None, None),
+        };
+
+        Self {
+            router_handle,
+            rx,
+            tx,
+            peer,
+            internal_addr,
+            rx_addr: None,
+        }
+    }
+
+    pub(crate) fn internal_addr(&self) -> &Address {
+        &self.internal_addr
+    }
+
+    /// Create a ([`UdsSendWorker`],[`WorkerPair`]) without spawning the worker.
+    pub(crate) async fn new_pair(
+        router_handle: UdsRouterHandle,
+        stream: Option<UnixStream>,
+        peer: SocketAddr,
+        pathnames: Vec<String>,
+    ) -> Result<(Self, WorkerPair)> {
+        let tx_addr = Address::random_tagged("UdsSendWorker_tx_addr");
+        let int_addr = Address::random_tagged("UdsSendWorker_int_addr");
+        let sender = UdsSendWorker::new(router_handle, stream, peer.clone(), int_addr);
+        Ok((
+            sender,
+            WorkerPair {
+                paths: pathnames,
+                peer,
+                tx_addr,
+            },
+        ))
+    }
+
+    /// Create a ([`UdsSendWorker`],[`WorkerPair`]) while spawning and starting the worker.
+    pub(crate) async fn start_pair(
+        ctx: &Context,
+        router_handle: UdsRouterHandle,
+        stream: Option<UnixStream>,
+        peer: SocketAddr,
+        hostnames: Vec<String>,
+    ) -> Result<WorkerPair> {
+        trace!("Creating new UDS worker pair");
+        let (mut worker, pair) = Self::new_pair(router_handle, stream, peer, hostnames).await?;
+
+        // TODO: @uds (Oakley) determine why his is considered bad in equivalent usage found here:
```

Comment:
Tcp crate now generates `rx_addr` in `new_pair`, this way there is only one place where this address is generated, which leads to a little bit cleaner code. Could you please also make that change for uds?